GCC 16 array subscript out-of-bounds warning

**Describe the bug**

Using GCC 16 optimization and enabling warnings generates a (false) warning about array subscript out-of-bounds.

Although the warning is not obviously valid, it would be great to have it addressed.

I used the following file test.cpp

```
#include <iostream>
#include "jsoncons/json.hpp"

using namespace jsoncons;

int main() {
    // Parse a JSON string
    std::string json_string = R"({"name": "John Doe", "age": 30})";
    json val = json::parse(json_string);

    // Access and print the values
    std::cout << "Name: " << val["name"].as_string() << "\n";
    //std::cout << "Age: " << val["age"].as_integer() << "\n";

    return 0;
}
```

in the gcc:16 docker file

```
docker run --rm -v "$(pwd)":/usr/src/app -w /usr/src/app gcc:16 g++ -I./include -O3 -Wall -pedantic test.cpp
```

resulting in

```
In file included from ./include/jsoncons/json_exception.hpp:18,
                 from ./include/jsoncons/conv_error.hpp:17,
                 from ./include/jsoncons/basic_json.hpp:31,
                 from ./include/jsoncons/json.hpp:10,
                 from test.cpp:2:
In function 'typename std::enable_if<(jsoncons::ext_traits::is_char8<CharT>::value && jsoncons::ext_traits::is_char32<CodepointT>::value), jsoncons::unicode_traits::convert_result<CharT> >::type jsoncons::unicode_traits::to_codepoint(const CharT*, const CharT*, CodepointT&, conv_flags) [with CharT = char; CodepointT = unsigned int]',
    inlined from 'std::size_t jsoncons::detail::escape_string(const CharT*, std::size_t, bool, bool, Sink&) [with CharT = char; Sink = jsoncons::string_sink<std::__cxx11::basic_string<char> >]' at ./include/jsoncons/json_encoders.hpp:90:58:
./include/jsoncons/utility/unicode_traits.hpp:431:52: warning: array subscript 6 is above array bounds of 'const uint32_t [6]' {aka 'const unsigned int [6]'} [-Warray-bounds=]
  431 |         ch -= offsets_from_utf8[extra_bytes_to_read];
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
./include/jsoncons/utility/unicode_traits.hpp: In function 'std::size_t jsoncons::detail::escape_string(const CharT*, std::size_t, bool, bool, Sink&) [with CharT = char; Sink = jsoncons::string_sink<std::__cxx11::basic_string<char> >]':
./include/jsoncons/utility/unicode_traits.hpp:152:20: note: while referencing 'jsoncons::unicode_traits::offsets_from_utf8'
  152 |     const uint32_t offsets_from_utf8[6] = { 0x00000000UL, 0x00003080UL, 0x000E2080UL,
      |                    ^~~~~~~~~~~~~~~~~
```

- Compiler:  g++ (GCC) 16.1.0
- Architecture (e.g. x86, x64) ____
- Operating system: ________

**What jsoncons library version?**

 - [ ] Latest release 1.7.0
 - [ ] Other release ______
 - [x] master

A work-around is to add a single element to the array

```
% git diff --ignore-space-change 
diff --git a/include/jsoncons/utility/unicode_traits.hpp b/include/jsoncons/utility/unicode_traits.hpp
index 33c03eabf..f874f08a2 100644
--- a/include/jsoncons/utility/unicode_traits.hpp
+++ b/include/jsoncons/utility/unicode_traits.hpp
@@ -149,8 +149,8 @@ namespace unicode_traits {
      * This table contains as many values as there might be trailing bytes
      * in a UTF-8 sequence. Source: ConvertUTF.c
      */
-    const uint32_t offsets_from_utf8[6] = { 0x00000000UL, 0x00003080UL, 0x000E2080UL, 
-                  0x03C82080UL, 0xFA082080UL, 0x82082080UL };
+    const uint32_t offsets_from_utf8[7] = { 0x00000000UL, 0x00003080UL, 0x000E2080UL,
+                  0x03C82080UL, 0xFA082080UL, 0x82082080UL, 0x00000000UL };
 
     /*
      * Once the bits are split out into bytes of UTF-8, this is a mask OR-ed
@@ -1331,4 +1331,3 @@ namespace unicode_traits {
 } // namespace jsoncons
 
 #endif //JSONCONS_UTILITY_UNICODE_TRAITS_HPP
-
```

But another option used elsewhere in the package is to suppress the specific compiler warning at the offending line.

--- a/include/jsoncons/utility/unicode_traits.hpp
+++ b/include/jsoncons/utility/unicode_traits.hpp
@@ -148,9 +148,11 @@ namespace unicode_traits {
      * Magic values subtracted from a buffer value during UTF8 conversion.
      * This table contains as many values as there might be trailing bytes
      * in a UTF-8 sequence. Source: ConvertUTF.c
+     *
+     * Seventh element added to satisfy GCC 16; see issue #712
      */
-    const uint32_t offsets_from_utf8[6] = { 0x00000000UL, 0x00003080UL, 0x000E2080UL, 
-                  0x03C82080UL, 0xFA082080UL, 0x82082080UL };
+    const uint32_t offsets_from_utf8[7] = { 0x00000000UL, 0x00003080UL, 0x000E2080UL,
+                  0x03C82080UL, 0xFA082080UL, 0x82082080UL, 0x00000000UL };
 
     /*
      * Once the bits are split out into bytes of UTF-8, this is a mask OR-ed

